src/leaderboard/leaderboard_resource.rs

```diff
@@ -0,0 +1,128 @@
+use crate::leaderboard::messages::data::fetch::FetchResponse;
+use crate::leaderboard::messages::data::fetch_by_rank::{FetchByRankRequest, RankRange};
+use crate::leaderboard::messages::data::fetch_by_score::{FetchByScoreRequest, ScoreRange};
+use crate::leaderboard::messages::data::get_rank::{GetRankRequest, GetRankResponse};
+use crate::leaderboard::messages::data::length::{LengthRequest, LengthResponse};
+use crate::leaderboard::messages::data::remove_elements::{
+    RemoveElementsRequest, RemoveElementsResponse,
+};
+use crate::leaderboard::messages::data::upsert::{IntoElements, UpsertRequest, UpsertResponse};
+use crate::leaderboard::MomentoRequest;
+use crate::MomentoResult;
+
+use momento_protos::leaderboard::leaderboard_client::LeaderboardClient as SLbClient;
+use tonic::codegen::InterceptedService;
+use tonic::transport::Channel;
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+
+static NEXT_DATA_CLIENT_INDEX: AtomicUsize = AtomicUsize::new(0);
+
+use super::messages::data::IntoIds;
+
+/// Represents a remote leaderboard resource.
+pub struct Leaderboard {
+    data_clients: Vec<SLbClient<InterceptedService<Channel, HeaderInterceptor>>>,
+    deadline: Duration,
+    cache_name: String,
+    leaderboard_name: String,
+}
+
+impl Leaderboard {
+    /// Delete a leaderboard.
+    pub async fn delete(&self) -> MomentoResult<DeleteResponse> {
+        let request = DeleteRequest::new();
+        request.send(self).await
+    }
+
+    /// Fetch elements from a leaderboard by rank.
+    pub async fn fetch_by_rank(
+        &self,
+        rank_range: impl Into<RankRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByRankRequest::new(rank_range);
+        request.send(self).await
+    }
+
+    /// Get elements from a leaderboard by score.
+    pub async fn fetch_by_score(
+        &self,
+        score_range: impl Into<ScoreRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByScoreRequest::new(score_range);
+        request.send(self).await
+    }
+
+    /// Get the length of a leaderboard.
+    pub async fn length(&self) -> MomentoResult<LengthResponse> {
+        let request = LengthRequest::new();
+        request.send(self).await
+    }
+
+    /// Get rank of elements from a leaderboard using their element ids.
+    pub async fn get_rank<T: IntoIds>(&self, ids: T) -> MomentoResult<GetRankResponse> {
+        let request = GetRankRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Remove elements from a leaderboard using their element ids.
+    pub async fn remove_elements<T: IntoIds>(
+        &self,
+        ids: T,
+    ) -> MomentoResult<RemoveElementsResponse> {
+        let request = RemoveElementsRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Upsert (update/insert) elements into a leaderboard.
+    pub async fn upsert<E: IntoElements>(&self, elements: E) -> MomentoResult<UpsertResponse> {
```

Comment:
Similar comments as above.

For the trait though, did you consider not creating IntoElements and instead using the usual Rust `Into` on a single value? You would delete IntoElements, write `impl From<(u32, f64)> for Element` and then use simple functions:
`pub async fn upsert(&self, element: impl Into<Element>)`
`pub async fn upsert_batch(&self, elements: impl IntoIterator<Item = impl Into<Element>>)`

This would also enable your users to more simply implement translators from their own business objects into unary and batch compatible representations.

Reply:
I've made the change to use `impl IntoIterator` in [981b5b7](https://github.com/momentohq/client-sdk-rust/pull/438/commits/981b5b7a02f5a5bb5acc27bca5ec941914e409d1). I decided to leave this as a single `upsert` function for now to align with the leaderboard pattern in other languages.